### PR TITLE
Add new RSS Block

### LINKED ISSUE #1799: Add Block: RSS
## Attributes

- RSS Feed URL
- Number of posts to show
- Display content
- Display author if available
- Display date

## States

Selected:
rss
rss feed with dates
rss feed with author
rss feed with dates and author
rss feed with dates author and excerpt

Neutral:

rss block

-----

Questions:
- Do we need to display the feed’s site title in the block itself? I didn't include this in the mockup.
- Can we get a post’s featured image from RSS?
- Can we display RSS posts as a grid? (Does this make a lot of sense if we can’t get featured images?)

### PR DESCRIPTION
**Pull request by Soean**
## Description
New Block to display RSS Feeds.

Fixes #1799

## Features:
- Add Feed URL
- Select number of items
- Display author
- Display date
- Display excerpt
- Select except length
- List and column layout
- Select column count

## Screenshots
jul-15-2018 13-02-58

Placeholder
bildschirmfoto 2018-07-15 um 13 14 11

Add URL `https://wordpress.org/news/`
bildschirmfoto 2018-07-15 um 13 14 52

Default settings
bildschirmfoto 2018-07-15 um 13 15 04

Enable date, author and except (20 words)
bildschirmfoto 2018-07-15 um 13 15 40

Enable grid layout
bildschirmfoto 2018-07-15 um 13 15 57

Settings with grid layout
bildschirmfoto 2018-07-15 um 13 15 53

Parsing error (Only in backend)
bildschirmfoto 2018-07-15 um 14 05 44

### CODE REVIEW

File: core-blocks/rss/index.php
Comment (Soean) [inline core-blocks/rss/index.php]: phpcs: `Silencing errors is discouraged`

**Comment by karmatosed** [comment]
@Soean could you refresh this please so a design review can happen?

**Comment by Soean** [comment]
@karmatosed The PR is refreshed and the design review can happen

**Comment by gziolo** [comment]
@Soean, can you rebase this PR again? For some reasons, PHP unit tests are failing. I have a review of this PR scheduled for later this week. @ajitbohra, not sure if you plan to help with the review, but I think that we might want to start with this one as @Soean can't fully self-review his work :)

File: packages/block-library/src/rss/index.php
Comment (Soean) [inline packages/block-library/src/rss/index.php]: @gziolo This line is causing the PHP waring:
````
Silencing errors is discouraged
Generic.PHP.NoSilencedErrors.Discouraged)
````

I copied this function from the widget:
https://github.com/WordPress/wordpress-develop/blob/6b366c6620fdd5960cedcdf80955966a715efa82/src/wp-includes/widgets/class-wp-widget-rss.php#L70

**Comment by ajitbohra** [comment]
@gziolo yes PR test/review on it :)

File: packages/block-library/src/rss/index.php
Comment (noisysocks) [reply packages/block-library/src/rss/index.php]: I can't think of any reason we ought to be suppressing errors here. Best to just remove the `@`, I think.

**Comment by noisysocks** [comment]
Hey @Soean! Thanks so much for all your hard work on this, it's looking great! Leaving you some quick comments on the UX and functionality before I dive into a code review.

#### 1. Placeholder state

Here's how the RSS placeholder state looks in this PR:

localhost_8888_wp-admin_post php_post 28 action edit desktop

This matches the placeholder state of the Embed block:

localhost_8888_wp-admin_post php_post 28 action edit desktop 1

We have a different style of placeholder state for the Shortcode block:

localhost_8888_wp-admin_post php_post 28 action edit desktop 2

For reference, here's how the RSS placeholder state looks in the design per #1799:



I'm wondering: Should the RSS block's placeholder state look more like the Shortcode block's placeholder state as per the design in #1799? cc. @melchoyce 

#### 2. Error state

Here's what an error looks like:

localhost_8888_wp-admin_post php_post 28 action edit desktop 3

There's a few problems here:

1. The error message should appear above the URL text field and _Use URL_ button so that the user is able to re-enter a valid URL right away.
2. The error message box needs some top and bottom padding. It's likely that the text just needs to be wrapped with a `<p>`.
3. The inspector controls in the sidebar (e.g. _Display author_) and the toolbar controls shown above (e.g. _Grid View_) should not be shown when the RSS block is in an error state.

For reference, her's what an error looks like in the Image block:

localhost_8888_wp-admin_post php_post 28 action edit desktop 4

#### 3. By-lines

When both _Display author_ and _Display date_ are enabled, the by-line under each RSS entry looks like this:

localhost_8888_wp-admin_post php_post 28 action edit desktop 5

Note that the by-line is:

```
January 22, 2019   By John Gruber
```

It should be:

```
January 22, 2019 by John Gruber
```

For reference, here's how this looked in the design per #1799:



#### 4. Excerpt length

Being able to control the except length is a really nice touch! Thanks for adding it.

screen shot 2019-01-23 at 11 44 40


1. Does it make sense that the minimum excerpt length is 0? When would a excerpt length of 0 be useful?
2. Why does pressing the Up/Down stepper arrows on the text field increment/decrement the excerpt length by 5? I would expect to be able to control this more finely. That is, I'd expect to be able to increment/decrement the text field by 1 using the stepper.

#### 5. PHP warnings

When I preview a post that contains an RSS block, I'm seeing lots of PHP warnings:

localhost_8888__p 28 preview_id 28 preview_nonce a6ecba3f44 preview true desktop

**Comment by Soean** [comment]
@noisysocks 
Thanks for your review.

1. **Placeholder**: The block needs a button and the default preview needs more space. This is why I  have chosen the design like the embed block.

2. **Error**: Thanks good point, i will work on it.

3. **By-lines**: Thanks, i fixed it.

4. **Excerpt**: If you use 0 as minimum you see just `[...]`.  It is not that useful, but is 1 word as minimum better? 
The stepper is now 1 word.

5. **Warning**: This is a waring from SimplePie for some RSS feeds. We have the same error in the widget.
>SimplePie is a third-party library and core isn't using the latest one (see #36669). The warning has already been fixed in https://github.com/simplepie/simplepie/commit/1111dab1baced14aaf05509c41ee416c3fc6a961.

source https://core.trac.wordpress.org/ticket/42515

WordPress 5.2 will update the library: https://core.trac.wordpress.org/ticket/36669

**Comment by ajitbohra** [comment]
While fetching RSS `BlockControls` & `InspectorControls` are avilable

processing

**Comment by ajitbohra** [comment]
@Soean Tested a few other blocks using `ServerSideRender` same behavior, seems like time to revisit `ServerSideRender` to know if the block is being rendered, in processing or an error occurred.

**Review by ajitbohra — APPROVED** [review]
Code wise LGTM :+1:  flagging for the second review

File: test/integration/full-content/fixtures/core__rss.serialized.html
Comment (gziolo) [inline test/integration/full-content/fixtures/core__rss.serialized.html]: I'm personally very surprised seeing that it doesn't contain any metadata in HTML comment. It looks like it's the same case for other blocks like Latest Posts. I will investigate it separately.

File: packages/block-library/src/rss/index.php
Comment (gziolo) [inline packages/block-library/src/rss/index.php]: My PHP is a bit rusty, but I bet this check should be performed on `$attributes['feedURL']` to prevent notices. It doesn't look like there is a default value provided for this property. Unless it's handled somehow in the code.

File: packages/block-library/src/rss/index.php
Comment (gziolo) [inline packages/block-library/src/rss/index.php]: What does this loop do? Can you explain? It might be worth adding code comment. I see the same code in core 🤷‍♂️

File: packages/block-library/src/rss/index.php
Comment (gziolo) [inline packages/block-library/src/rss/index.php]: It might make sense to use const to share it with attributes. It took me a while to understand why `5` was picked.

File: packages/block-library/src/rss/edit.js
Comment (gziolo) [inline packages/block-library/src/rss/edit.js]: Note to self. Check in other places whether we have to use `IconButton` with all those class names listed. It might mean we could expose `ToolbarButton` component to simplify it.

File: packages/block-library/src/rss/edit.js
Comment (gziolo) [inline packages/block-library/src/rss/edit.js]: Out of curiosity, why `Edit RSS URL` can't be handled the same as `layoutControls`? I think it's similar in other places, but I'm willing to learn why :)

**Review by gziolo — COMMENTED** [review]
I finished my first pass. I need to spend some more time on PHP file but I'm not an expert in that. I would appreciate some help from WordPress core contributors.

**Comment by gziolo** [comment]
> Tested a few other blocks using ServerSideRender same behavior.

@ajitbohra & @Soean let's tackle it separately. It's a minor glitch which is completely unrelated to this PR.

**Comment by gziolo** [comment]
@mapk and @melchoyce would you mind doing design review. Code wise it looks very close to be ready, we reached nitpicking phase :)

File: packages/block-library/src/rss/index.php
Comment (gziolo) [inline packages/block-library/src/rss/index.php]: Looking at screenshots in the description, I started wondering whether we should hide that very developer oriented messages from the users or at least warn about its technical nature :)

**Comment by melchoyce** [comment]
I know the other error @gziolo was the same, but we shouldn't be using the `.notice-alt` class here, since that class specifically exists to be displayed on a white background. Since the placeholder has a grey background, we should be using the regular notice class:

image

Are we able to rewrite this error message to something more helpful?

Also, would be good to retain the input field and display the error above, so you can go and edit the URL you entered and try again. As it is now, you need to delete the block and re-add it in order to recover.

File: packages/block-library/src/rss/style.scss
Comment (Soean) [reply packages/block-library/src/rss/style.scss]: I replaced the values with variables, now it is nearly the same as `latest-posts`. Maybe we should generalize it?

File: packages/block-library/src/rss/index.php
Comment (Soean) [reply packages/block-library/src/rss/index.php]: It trims the string until it starts with `http`. I don't think we need this here anymore, because we have other checks in PHP now.

File: packages/block-library/src/rss/editor.scss
Comment (Soean) [reply packages/block-library/src/rss/editor.scss]: I changed it to `.block-editor`

File: packages/block-library/src/rss/index.js
Comment (Soean) [reply packages/block-library/src/rss/index.js]: Added `feed` and `atom` as keywords.

File: lib/load.php
Comment (Soean) [reply lib/load.php]: done

File: packages/block-library/src/rss/index.php
Comment (swissspidy) [reply packages/block-library/src/rss/index.php]: That is correct, yes. No text domain for that.

Also, I‘d probably use sprintf for the block name here and use double quotes for the string in order to avoid backslashes for the single quotes in it.

File: packages/block-library/src/rss/index.php
Comment (swissspidy) [reply packages/block-library/src/rss/index.php]: Why is this being done? Doesn‘t really make sense to me. If the URL is invalid after just trimming, it‘s invalid.

File: packages/block-library/src/rss/index.php
Comment (swissspidy) [inline packages/block-library/src/rss/index.php]: Can‘t use filter_var because it might not be available. Use sanitization and escaping functions built into core of needed. I don‘t think it‘s necessary here at all. fetch_feed() should handle that already, so I think you should just remove this line

File: packages/block-library/src/rss/index.php
Comment (swissspidy) [inline packages/block-library/src/rss/index.php]: I feel like this extra function is overkill, especially because it is in global namespace. Just add this markup inline.

File: packages/block-library/src/rss/index.php
Comment (swissspidy) [inline packages/block-library/src/rss/index.php]: I think this check/siggestion should only be in the editor, not in the render callback. I‘d remove it here.

File: packages/block-library/src/rss/index.php
Comment (Soean) [reply packages/block-library/src/rss/index.php]: I removed this line and will do some tests with different inputs.

File: packages/block-library/src/rss/edit.js
Comment (Soean) [reply packages/block-library/src/rss/edit.js]: Thanks, this is way better.

File: packages/block-library/src/rss/index.php
Comment (swissspidy) [reply packages/block-library/src/rss/index.php]: I'd suggest use `wp_parse_args()` at the beginning of the function to set some defaults there as well.

File: packages/block-library/src/rss/index.php
Comment (swissspidy) [inline packages/block-library/src/rss/index.php]: Why the `strip_tags`? Shouldn't `esc_url` be enough already?

Comment (Soean) [reply packages/block-library/src/rss/index.php]: Should be enough. This line is from the rss widget.

**Review by mapk — DISMISSED** [review]
Just tested and I see the same error problem that @melchoyce caught above. Looking at the Cover block to see how this gets resolved.

cover-error

That lead me to this mockup if those are the right patterns:

error-rss

Would be great to keep the user on the same screen as the input field when the error displays. Thanks for working on this!

File: packages/block-library/src/rss/index.php
Comment (noisysocks) [inline packages/block-library/src/rss/index.php]: How do you feel about combining the author and date into a single sentence and capitalising the `b` in `by` when the author is enabled but the date is disabled?

```php
$date = '';
if ( $attributes['displayDate'] ) {
	$date = $item->get_date( 'U' );

	if ( $date ) {
		$date = sprintf(
			'<time datetime="%1$s">%2$s</time> ',
			date_i18n( get_option( 'c' ), $date ),
			date_i18n( get_option( 'date_format' ), $date )
		);
	}
}

$author = '';
if ( $attributes['displayAuthor'] ) {
	$author = $item->get_author();
	if ( is_object( $author ) ) {
		$author = $author->get_name();
		$author = esc_html( strip_tags( $author ) );
	}
}

$byline = '';
if ( $date && $author ) {
	$byline = '<span class="wp-block-rss__byline>' . sprintf( __( '%s by %s' ), $date, $author ) . '</span>';
} elseif ( $date ) {
	$byline = '<span class="wp-block-rss__byline>' . $date . '</span>';
} elseif( $author ) {
	$byline = '<span class="wp-block-rss__byline>' . sprintf( __( 'By %s' ), $author ) . '</span>';
}
```

File: packages/block-library/src/rss/index.php
Comment (noisysocks) [inline packages/block-library/src/rss/index.php]: Same as above: No need to call `__destruct()` or `unset()` here.

**Review by noisysocks — COMMENTED** [review]
This tests well and is looking really good! Great job on pushing this along ❤️

I left a few suggestions—nothing major. I'd be OK with this landing as is and iterating.

**Comment by noisysocks** [comment]
> Would be great to keep the user on the same screen as the input field when the error displays. Thanks for working on this!

Unfortunately it's difficult to do this because that error is generated by the server and displayed on the client using `ServerSideRender`. The client isn't able to distinguish between a successful response and an error response. 

I suggest landing this and opening a bug report for the error message issue since it will require changes to `ServerSideRender` or for the RSS block to move away from `ServerSideRender` entirely.

**Comment by jasmussen** [comment]
Holy guacamole this is such an amazing block. TOO COOL! 

GIF-time:

rssblock

SHIP IT SO FAST OUR HEADS SPIN.

I agree with both @mapk and @noisysocks — it would be good to polish up how the error messages show up, but let's do that separately from this PR.

File: packages/block-library/src/rss/index.php
Comment (Soean) [reply packages/block-library/src/rss/index.php]: This lines were added because of memory leaks in older PHP versions. It was fixed in PHP 5.3: http://simplepie.org/wiki/faq/i_m_getting_memory_leaks

WordPress is increasing its minimum in WP 5.2 to PHP 5.6. Maybe we should do the same, as all new code is for WP 5.2.

I removed the unnecessary code.

Comment (Soean) [reply packages/block-library/src/rss/index.php]: I opened a ticket for bumping the PHP minimum: https://github.com/WordPress/gutenberg/issues/13478

Comment (Soean) [reply packages/block-library/src/rss/index.php]: Yes sure, this are only 4 extra lines. I added them back with a comment about PHP 5.2 compatibility

Comment (youknowriad) [reply packages/block-library/src/rss/index.php]: Thanks, it looks like this PR is very close to land.

**Comment by youknowriad** [comment]
Looks like there's some small php code style failures

**Comment by Soean** [comment]
Already fixed

File: packages/block-library/src/rss/index.php
Comment (gziolo) [inline packages/block-library/src/rss/index.php]: This needs to be WordPress the same as here:
https://github.com/WordPress/gutenberg/blob/master/packages/block-library/src/latest-posts/index.php#L5

It's not clear but this is because it gets copied to core, it's similar to translations. I'm wondering where we should put such guidelines. I wouldn't know about it if I didn't see changes in the core after Grunt run a few weeks back :)

Comment (Soean) [reply packages/block-library/src/rss/index.php]: Done

**Review by gziolo — APPROVED** [review]
Code wise it looks good. All my comments were addressed. Great work @Soean on RSS block. Thanks for updating PR instantly after all feedback received.

@mapk there is agreement that error state needs to get some improvement. There are some technical challenges which we need to look at first to get it done. Given that we still have more than a week until next planned Gutenberg release, I would suggest we merge this PR in the current shape and improve UI separately.

**Comment by gziolo** [comment]
There is also possible improvements to tackle in the follow-up PR.

1. Set the minimum number of words to 1 or something higher. 0 doesn't look good:

screen shot 2019-01-24 at 14 06 42

2. It would be good to clarify that the length of excerpt is the number of words displayed:

screen shot 2019-01-24 at 14 08 15

3. When there is an error we still display all controls (it's related to the aforementioned change to the way `ServerSideRender` component works:

screen shot 2019-01-24 at 14 09 36

**Comment by melchoyce** [comment]
I'm cool with addressing the error messages in a follow-up PR. Can we get a list of all the strings in that future PR?

**Comment by youknowriad** [comment]
🎉 Awesome work everyone involved here. Let's get this in and iterate.

**Comment by gziolo** [comment]
I collected all issues raised in #13498. Feel free to update if I missed anything.

**Comment by markzahra** [comment]
Hey folks, not sure if this is the right place to ask (if not, let me know where would be best)...

What was the main reason for looking into adding this block to WordPress Core back in 2017?

I ask as the leader of the current team behind [WP RSS Aggregator](https://wordpress.org/plugins/wp-rss-aggregator/). I just came across this issue as we've been working on introducing our own Gutenberg block. Essentially, our free core version's block does exactly the same thing as this block, then the rest of our features and add-ons take it a step or two further.

I'm wondering what the future holds for this Gutenberg block to see if and how it may impact our plugin going forward. I've read the related Github issues but couldn't find what actually created the need for this feature to be in WordPress core. It would be great if I could hear some feedback on that and the future plans for the RSS block.

**Comment by melchoyce** [comment]
Hey @markzahra, this block is part of an effort to [convert all existing core widgets into blocks](https://make.wordpress.org/core/2018/12/08/9-priorities-for-2019/). You can read more about that here: https://make.wordpress.org/core/2018/12/17/status-update-porting-widgets-to-blocks/

I don't think we'll do much beyond what's in this PR, except for maintenance.

File: test/integration/full-content/fixtures/core__rss.serialized.html
Comment (aduth) [reply test/integration/full-content/fixtures/core__rss.serialized.html]: > I'm personally very surprised seeing that it doesn't contain any metadata in HTML comment. It looks like it's the same case for other blocks like Latest Posts. I will investigate it separately.

It is surprising, as it would imply that given the input of `core__rss.html`, the editor would destroy all attributes in a reserialization.

There are also a number of attributes defined in `core__rss.html` which simply don't exist for the defined block type (e.g. `postsToShow`), perhaps part of earlier iterations of the block.

Fix proposed at #14122

**Pull request merged**
